File: c/grains/grains.c
#include "grains.h"

#define SQUARE_MIN 1
#define SQUARE_MAX 64

/**
 * Grains
 * https://exercism.org/tracks/c/exercises/grains
 */

uint64_t square(uint8_t index)
{
    if (index < SQUARE_MIN || index > SQUARE_MAX) {
        return 0;
    }

    uint64_t res = 1;
    for (int i = SQUARE_MIN; i < index; i++) {
        res *= 2;
    }

    return res;
}

uint64_t total(void)
{
    uint64_t res = 0;
    uint64_t n = 1;
    for (int i = SQUARE_MIN; i <= SQUARE_MAX; i++) {
        res += n;
        n *= 2;
    }

    return res;
}

File: java/grains/Grains.java
import java.math.BigInteger;

/**
 * Grains
 * https://exercism.org/tracks/java/exercises/grains
 */
class Grains {

    private static final BigInteger TWO = BigInteger.valueOf(2);
    private static final int SQUARE_COUNT = 64;

    BigInteger grainsOnSquare(final int square) {
        if (square < 1 || square > SQUARE_COUNT) {
            throw new IllegalArgumentException("square must be between 1 and " + SQUARE_COUNT);
        }

        return TWO.pow(square - 1);
    }

    BigInteger grainsOnBoard() {
        BigInteger grainsCount = BigInteger.ZERO;
        for (int square = 1; square <= SQUARE_COUNT; square++) {
            grainsCount = grainsCount.add(grainsOnSquare(square));
        }

        return grainsCount;
    }
}
